fmt: Canonicalize symbol names

hryx opened the pull request:
Fixes #166

## Overview

This adds new behavior to `zig fmt` which normalizes (renders a canonical form of) quoted identifiers like `@"hi_mom"` to some extent. This can make codebases more consistent and searchable.

To prevent making the details of Unicode and UTF-8 dependencies of the Zig language, only bytes in the ASCII range are interpreted and normalized. Besides avoiding complexity, this means invalid UTF-8 strings cannot break `zig fmt`.

Both the tokenizer and the new formatting logic may overlook certain errors in quoted identifiers, such as nonsense escape sequences like `\m`. For now, those are ignored and we defer to existing later analysis to catch.

This change is not expected to break any existing code.

## Behavior

Below, "verbatim" means "as it was in the original source"; in other words, not altered.

- If an identifier is bare (not quoted) in the original source, no change is made. Everything below only applies to quoted identifiers.
- Quoted identifiers are processed byte-wise, not interpreted as a UTF-8 sequence.
- `\x` and `\u` escapes are processed.
    - If the escape sequence itself is invalid, the sequence is rendered verbatim.
    - If the un-escaped codepoint is not ASCII, the sequence is rendered verbatim.
    - Otherwise, the character is rendered with [`formatEscapes`](https://github.com/ziglang/zig/blob/cbdd21c/lib/std/zig/fmt.zig#L50-L87) rules: either literally or `\x`-escaped as needed.
- If the resulting identifier still contains escapes, it it remains quoted.
- If the resulting identifier is not a valid bare symbol (`[A-Za-z_][A-Za-z0-9_]*`), it remains quoted.
- If the resulting identifier is a keyword, it remains quoted.
- If the resulting identifier is a primitive value/type (`i33`, `void`, `type`, `null`, `_`, etc.), it is rendered unquoted if the syntactical context allows it (field access, container member), otherwise it remains quoted.
- Otherwise, it is unquoted. Celebrate in the manner of your choosing.

ifreund commented on lib/std/zig/render.zig:
```diff
@@ -2377,6 +2379,28 @@ fn renderSpace(ais: *Ais, tree: Ast, token_index: Ast.TokenIndex, lexeme_len: us
 
+// If the symbol is escaped with @"" syntax, only render it that way if it
+// contains invalid characters, otherwise render it with canonical syntax.
+fn renderCanonicalSymbol(gpa: Allocator, ais: *Ais, tree: Ast, token_index: Ast.TokenIndex, space: Space) Error!void {
+    const lexeme = tokenSliceForRender(tree, token_index);
+    if (lexeme[0] == '@') {
+        const symbol = std.zig.string_literal.parseAlloc(gpa, lexeme[1..]) catch |err| switch (err) {
+            error.InvalidLiteral => unreachable,
+            else => |e| return e,
+        };
+        defer gpa.free(symbol);
+        if (std.zig.isValidId(symbol)) {
+            // Canonicalize symbol that was unnecessarily escaped with @""
+            try ais.writer().writeAll(symbol);
+        } else {
+            try ais.writer().writeAll(lexeme);
+        }
+    } else {
+        try ais.writer().writeAll(lexeme);
+    }
+    try renderSpace(ais, tree, token_index, lexeme.len, space);
+}
+
```
Hmm, I wonder if we could just check in `renderToken()` if the token has tag `identifier` and apply this logic there if so. This would avoid needing to update all these call sites to call the new function and potentially forgetting one. It would also be more robust to future code churn.

ifreund commented:
> The allocator is used to parse symbols escaped with @"", but there is probably a more efficient approach. I was thinking of adding something like this to std/zig/string_literal.zig:

Hmm, yeah there should be no need to allocate here. inlining the logic of `std.zig.isValidId()` into the parsing would avoid the need to do any extra in-memory copy in the case where nothing needs to be changed. The function signature you suggest that takes a writer would solve the other case when the output of `zig fmt` should differ from the input as an identifiers needs to be canonicalized.

hryx commented:
Thanks @ifreund, great suggestions. Agreed about guarding against future churn. My original thinking was to avoid duplicating the code from `isValidId()`, but now I think it makes more sense to combine the parsing and "validation" logic here.

hryx commented:
The suggestion worked like a charm. Now it covers everything, doesn't allocate, and is much cleaner.

Last concern: I duplicated most of the code from `parseAppend` into `parseWrite`. It's not so nice, but the tiny differences didn't make for an obvious (to me) abstraction. I can try my hand at improving code reuse if you'd like, thought I'd push what I have in case you already have a pattern in mind that would fit nicely here.

hryx commented:
CI caught a missed case: identifiers that are escaped keywords. Added a stub fix that currently fails but I can come back to look closer shortly. (Apologies for the noise)

hryx commented:
As mentioned by @InKryption [in a comment](https://github.com/zigtools/zls/pull/724#issuecomment-1294516558):

> fields with the name `type` can be written without `@""` syntax

This reminds me of some cases I hadn't considered:

```zig
const Pumpkin = struct {
    type: Squash, // define field name
    const Squash = struct {};
};

fn f() void {
    var p: Pumpkin = .{
        .type = .{}, // initializer
    };

    p.type = .{}; // field access
}
```

Identifiers spelled like keywords without `@""` escaping are legal in those cases. Should the above cases (maybe others I'm forgetting) be un-escaped?

ifreund commented:
> Identifiers spelled like keywords without @"" escaping are legal in those cases. Should the above cases (maybe others I'm forgetting) be un-escaped?

`type` isn't actually a keyword but rather an identifier in the global scope. Your code should actually already handle that example properly as `std.zig.isValidId()` returns true for `type`.

ifreund commented on lib/std/zig/render.zig:
```diff
@@ -2377,6 +2381,24 @@ fn renderSpace(ais: *Ais, tree: Ast, token_index: Ast.TokenIndex, lexeme_len: us
+    if (lexeme[0] == '@') {
+        assert(lexeme.len > 2 and lexeme[1] == '"' and lexeme[lexeme.len - 1] == '"');
+
+        const quoted_contents = lexeme[2 .. lexeme.len - 1];
+        if (std.zig.isValidId(quoted_contents)) {
```
Unfortunately this isn't sufficient to handle the example of `@"\x66\x6f\x6f"` given in the linked issue as it does not interpret escape sequences before checking `isValidId()`. `@"\x66\x6f\x6f"` should be canonicalized to `foo`.

ifreund replied:
It's possible in theory to handle this without allocation by combining the logic of `isValidId()` and `parseAppend()` into a single function with a fixed sized buffer large enough to hold the longest keyword.

Alternatively, it wouldn't be too bad to allocate for this if we reused the same buffer for each identifier. This would look something like a scratch `ArrayList(u8)` buffer that is created at the start of `renderTree()` and is only deinitialized when that function is completed. Every `renderCanonicalSymbol()` call would then parse the contents of `@""` quoted identifiers into it, check if the parsed result is a valid id, etc. Then the buffer would be cleared without freeing the memory so that the next `renderCanonicalSymbol()` call may avoid allocating at all if the buffer is large enough.

hryx replied:
Good ideas. After writing down my "Behavior of canonicalization" comment, I can visualize a two-pass approach that uses a local `var buf: [length_of_longest_keyword_or_primitive]u8`, which I agree is less nasty than option 2.

ifreund commented on lib/std/zig/string_literal.zig:
```diff
@@ -280,6 +280,43 @@ pub fn parseAlloc(allocator: std.mem.Allocator, bytes: []const u8) ParseError![]
+/// Parses `bytes` as a Zig string literal and writes the result to the std.io.Writer type.
+/// Asserts `bytes` has '"' at beginning and end.
+/// Higher level API. Does not return extra info about parse errors.
+pub fn parseWrite(writer: anytype, bytes: []const u8) ParseError!void {
```
I think that if this API returned a rich error like `parseAppend()` does having signature
```zig
pub fn parseAppend(writer: anytype, bytes: []const u8) error{OutOfMemory}!Result {
```
it would be strictly better and could replace the `parseAppend()` usage in `AstGen.zig` as well as the implementation of `parseAlloc()`. 

Note that `std.ArrayList(u8).writer()` exists, so the diff in `AstGen.zig` for example could be something like
```diff
diff --git a/src/AstGen.zig b/src/AstGen.zig
index 48e6a480f..0e2164681 100644
--- a/src/AstGen.zig
+++ b/src/AstGen.zig
@@ -9864,7 +9864,7 @@ fn parseStrLit(
 ) InnerError!void {
     const raw_string = bytes[offset..];
     var buf_managed = buf.toManaged(astgen.gpa);
-    const result = std.zig.string_literal.parseAppend(&buf_managed, raw_string);
+    const result = std.zig.string_literal.parseWrite(buf_managed.writer(), raw_string);
     buf.* = buf_managed.moveToUnmanaged();
     switch (try result) {
         .success => return,
```

If that improvement is made to this API I'd be happy to delete `parseAppend()` and remove the code duplication.

hryx commented:
> type isn't actually a keyword but rather an identifier in the global scope. Your code should actually already handle that example properly as std.zig.isValidId() returns true for type.

You are so right, it's not a keyword, I had a late-night brainfart. But there is still a problem as revealed by CI (which I'd hastily misinterpreted):

```
/usr/home/build/zig/zig-cache/o/e7f95f958298c1559827c567f0fba9c1/cimport.zig:669:24: error: name shadows primitive 'type'
pub inline fn CLITERAL(type: anytype) @TypeOf(type) {
                       ^~~~
/usr/home/build/zig/zig-cache/o/e7f95f958298c1559827c567f0fba9c1/cimport.zig:669:24: note: consider using @"type" to disambiguate
```

The note tells me that you can use `@""` to escape _primitive_ shadowing, is that correct? While local/global shadowing is never allowed.

Additionally, my previous comment had an incomplete example. Keeping `"@type"` quoted is necessary for vars and args (because of shadowing a primitive), while they can safely be un-escaped for struct fields.

The last complication here is:

```zig
fn f(@"type": bool) void {
    // These refer to different things:
    _ = @"type";
    _ = type;
}
```

hryx commented:
Because of the cases I previously overlooked, I thought I'd write out my assumptions below to make sure I'm targeting the right behavior.

The `\u -> \x` thing hasn't been mentioned anywhere as far as I know, so let me know if you agree with the proposal. I could see a case for doing it or not doing it, no strong opinion.

I think the un-escaping has an impact on our allocation discussion -- I'll have an update on that soon.

## Behavior of canonicalization

The following rules are applied in order to decide how to render an identifier token:

1. Identifiers **not** quoted with `@""` syntax are always written raw, and none of the below applies to them.
2. All `\x` and `\u` character escapes are unescaped when they match `[A-Za-z0-9_]`, regardless of the final name quoting.
3. All `\u` escapes that do not match basic pattern above, but fit in one byte, are normalized to `\x` escapes.
4. Invalid symbols (thoes which no not match `[A-Za-z_][A-Za-z0-9_]*`) are left quoted.
5. Keywords are left quoted.
6. Primitive types and values (`bool`, `type`, `_`, `true`, ...) are unquoted when it would be legal to do so (struct field, field access, enum literal, ...) and otherwise left quoted (variable, function name, arg, ...).
7. Everything else is unquoted.

ifreund commented:
> The note tells me that you can use @"" to escape primitive shadowing, is that correct? Whereas local/global shadowing is never allowed.

Ah yes, this is correct and makes this canonicalization a good bit more complicated unfortunately since it must be aware of where in the AST the identifier appears.

You're probably wondering how best to identify primitive types. The current answer to this is the `isPrimitive()` function in `src/AstGen.zig` in the self hosted compiler. I see that `translate-c` already imports this function from AstGen for it's own purposes. It probably makes sense to move it to `std.zig` since we need it here as well.

Those assumptions look great, thanks for taking the time to write them out explicitly :)

> This is opening up a bit is because it introduces normalization even when the thing remains quoted. I believe that fits the spirit of the original issue, but let me know if I'm reading too deeply into it!

This canonicalization from `@"\x65\x72\x72\x6F\x72"` to `@"error"` isn't something I had personally considered but it certainly fits the spirit of the issue and I think it would definitely be worth implementing.

This issue is turning out to be even more of a rabbit hole than I initially thought :D

hryx commented on lib/std/zig/render.zig:
```diff
@@ -2377,6 +2391,145 @@ fn renderSpace(ais: *Ais, tree: Ast, token_index: Ast.TokenIndex, lexeme_len: us
+        if (!std.zig.isValidId(buf[0..buf_i])) {
+            return renderQuotedIdentifier(ais, tree, token_index, space, false);
+        }
+        const isPrimitive = @import("../../../src/AstGen.zig").isPrimitive;
```
TODO: move this to std as suggested

hryx commented:
Snap, somehow my eyes glazed over this comment: https://github.com/ziglang/zig/blob/b40fc70188fd097e9e05b5c18b635b67b718380e/lib/std/zig/render.zig#L187-L188

To keep things simple, I'd be ok worrying about that till after this is PR is done since `renderSpace()` has no significant impact here. (On the other hand, the only thing I love more than writing code is deleting code, so...)

hryx commented:
> The current answer to this is the isPrimitive() function in src/AstGen.zig in the self hosted compiler. I see that translate-c already imports this function from AstGen for it's own purposes. It probably makes sense to move it to std.zig since we need it here as well.

It looks like the heart of `isPrimitive()` is the enum defined in `src/Zir.zig`, and I'm guessing it's pretty far-fetched to move ZIR into stdlib at this point. Maybe the enum itself can live in std, I just didn't see an obvious existing home for it.

What I'll do for tonight is copypaste the enum so as to get this change up to spec and get CI passing, then worry about organization tomorrow -- there's still cleanup to do anyway.

---

As a second matter, the secret final boss of rendering quoted identifiers is non-ASCII codepoints. I hadn't considered whether we should normalize e.g. `@"type\u{1f335}"` to `@"type🌵"`, or even vice versa.

I think I have a solve for now, but my dream solution is adopting the `%q` [format specifier from Go](https://pkg.go.dev/fmt). It normalizes strings more or less the way we want here. (Ignore superficial details like escape syntax)

Example: https://go.dev/play/p/ZZPRr5r9TS9

```go
func main() {
	fmt.Printf("Quoted by %%q specifier:               %q\n", "\u26a1 abc? ' \" 世界 cąbbäge \t 🐢")
	fmt.Printf("Quoted (ASCII-only) by %%+q specifier: %+q\n", "\u26a1 abc? ' \" 世界 cąbbäge \t 🐢")
	fmt.Printf("Quoted by %%#q specifier:              %#q\n", "\u26a1 abc? ' \" 世界 cąbbäge \t 🐢")
}
```

```
Quoted by %q specifier:               "⚡ abc? ' \" 世界 cąbbäge \t 🐢"
Quoted (ASCII-only) by %+q specifier: "\u26a1 abc? ' \" \u4e16\u754c c\u0105bb\u00e4ge \t \U0001f422"
Quoted by %#q specifier:              `⚡ abc? ' " 世界 cąbbäge 	 🐢`
```

It might not be a blocker for this PR though, let's see how the fix goes. I'd be very happy to propose or implement it.

hryx commented on lib/std/zig/render.zig:
```diff
@@ -2310,6 +2311,19 @@ fn renderTokenComma(ais: *Ais, tree: Ast, token: Ast.TokenIndex, space: Space) E
     }
 }
 
+/// Render an identifier, and the comma that follows it, if it is present in the source.
+/// If a comma is present, and `space` is `Space.comma`, render only a single comma.
+fn renderIdentifierComma(ais: *Ais, tree: Ast, token: Ast.TokenIndex, space: Space, quote: QuoteBehavior) Error!void {
+    const token_tags = tree.tokens.items(.tag);
+    const maybe_comma = token + 1;
+    if (token_tags[maybe_comma] == .comma and space != .comma) {
+        try renderIdentifier(ais, tree, token, .none, quote);
+        return renderToken(ais, tree, maybe_comma, space);
+    } else {
+        return renderIdentifier(ais, tree, token, space, quote);
+    }
+}
+
```
Copypasted from `renderTokenComma()` and adapted to identifiers. There's now only one usage of each so maybe a candidate for cleanup.

hryx commented:
CI error after formatting All The Code:

```
[ 98%] Built target zig1
[ 99%] Building stage2 object /Users/runner/work/1/s/build/zig2.o
./src/Zir.zig:2370:36: error: expected token ')', found '.'
                .ty = Type.initTag(.undefined),
                                   ^
```

My quick guess is there's a discrepancy in the stage1 and self-hosted tokenizers. I'll browse the C++, hopefully it's narrowly-scoped.

ifreund commented:
> My quick guess is there's a discrepancy in the stage1 and self-hosted tokenizers. I'll browse the C++, hopefully it's narrowly-scoped.

To be pedantic, the difference is in the parser not the tokenizer. In the self hosted compiler `true`, `false`, `null`, and `undefined` are primitives while in the C++ compiler they are still keywords. This was changed for stage2 in 05cf44933d753f7a5a53ab289ea60fd43761de57. I'm not sure what the least painful way to work around that discrepancy for the purposes of this PR would be, perhaps backporting that change to the C++ codebase.

hryx commented:
Ok, we had a chat and made some design conclusions and have a clear path to simplify/finish this PR.

Spec changes:

- The Zig language should not depend on the Unicode specification. Tossing Unicode out of the equation means we don't have to worry about invalid codepoints or UTF-8 sequences.
- We'll still parse `\x` and `\u` escapes, but will only un-escape them if they're ASCII and legal in a Zig string literal. Otherwise rendered verbatim. So if a source has both `@"🐢"` and `@"\u{01f422}"`, well, such is life.
- No `\u` -> `\x` simplification. Above `7f` that's irrelevant given the above. And even below that, they convey different things anyway, raw bytes vs.  codepoints.

Path to merge:

1. Back-port the keyword change to C++.
2. Implement above behavior changes.
3. Get CI passing.
4. Refine code style and reorganize stuff, including and stdlib changes.
5. Clean up commits and merge. :beers: 

Other notes:

- I uncovered that the tokenizer accepts nonsense escapes like `\m`, deferring the error to a later parsing stage. To keep the PR scoped, it's fine to keep that behavior in the new render code too.

hryx commented:
The render logic should be up to snuff now, so I'm currently stuck on getting CI to pass. It seems to be a different error every time. In my ignorance of the CI setup, I pushed after a fresh master rebase, so I hope that didn't blow away any useful clues.

Which commands should I learn to identify the issue locally? `zig build test-fmt` is clean. I tried `zig build test -Dskip-non-native`, but can't tell if the panic I get locally is a result of one of my changes (the C++ change?) or because I'm using the wrong test command.

hryx commented:
@ifreund This is now up to spec and ready for review! Here are some notes to keep in mind:

- I left the commit history a mess until you are satisfied with the changes themselves, but once you are, I will clean it up to something like this:
    1. Backport keyword change to stage1
    2. All functional changes
    3. Run zig fmt on the whole codebase
- I replaced `parseAppend` with `parseWrite` in std as discussed, and I like it, although it's not actually necessary for this PR anymore. Happy to remove the change or split it into its own commit if you prefer.
- See the later commit messages for a couple of unrelated side effects of the PR, namely https://github.com/ziglang/zig/pull/13343/commits/ead001200d8f5a3448b141de3bd246dc900ea3a2 (bug fix, you already know this one) and https://github.com/ziglang/zig/pull/13343/commits/a9a25aaa29fa13aba3c19105a74b7542b603a3ef.
- I'm not sure if I have properly tested the C++ change. I've been relying on CI and a zig-bootstrap'd stage3 (master) to test these updates. Let me know if you have concerns or tips on verifying that portion.

hryx commented:
Rebased on master and cleaned up. I'll keep this in a perpetual review-ready state from this point.

If you want to refer to any of the earlier commits that are relevant to previous discussion, I pushed the original branch to https://github.com/ziglang/zig/compare/master...hryx:zig:canonical-symbols-scratch

ifreund reviewed (approved):
LGTM as well! Thanks for your patience waiting for a merge. I'll just rebase this again locally and run zig fmt to make sure we don't break CI then merge it.
